### ISSUE TITLE
predict_generator may shuffle batches if used with workers != 1

### ISSUE THREAD

**Issue by GPhilo** [fchollet/keras#6745]
**Keras version**: 2.0.4 (backend is tensorflow, but it's irrelevant)

**Premise**: I have looked again and again at the library code and I *think* I am certain of this problem, but I might be missing something, so check my reasoning. I also, to the best of my googling skill, believe that the behaviour is undocumented.

**My use case**: I want to predict the labels of a big dataset of images and I want to associate each label with the corresponding filename. 

**My code**: 
```python
model = InceptionV3(weights='imagenet')
datagen = ImageDataGenerator(preprocessing_function=preprocess_input)
generator = datagen.flow_from_directory(dataset_dir,
                                        target_size=(299,299),
                                        batch_size=batch_size,
                                        class_mode=None,
                                        shuffle=False)
preds = model.predict_generator(generator, len(generator.filenames)//batch_size, workers=5)
```
Note how it's important that the order of the lables in `preds` matches the order of the filenames in `generator.filenames`. I want to use multiple workers to build the queue because with just one worker the GPU stalls between batches.

**The expected behaviour**: `predict_generator` spawns 5 worker threads that put batches in a queue **in the same order as they are produced by the generator**.

**The issue**: I believe there is a possibility for batch shuffling when batches are enqueued in `GeneratorEnqueuer::data_generator_task()`. Reshuffling happens, for example, in the following sequence of events.
In `data_generator_task()`:
1. `worker_1` calls `next(self._generator)` and gets `batch_1`, then gets suspended.
2. `worker_2` calls `next(self._generator)` and gets `batch_2`.
3. `worker_2` enqueues `batch_2`.
4. `worker_1` resumes execution and enqueues `batch_1`.
The two batches have been swapped, so the predictions do not match the corresponding filenames anymore.

**Comment by Dref360** [comment]
The behaviour of the thread scheduler is unpredictable so the notion of "order" is not possible. Overriding predict_generator to save the actual order would work.

**Comment by GPhilo** [comment]
If that's the case, what's the intended use of the multiple workers in predict_generator? If the order is not preserved, the labels produced cannot be matched to the data they come from.

I'm working [here](https://github.com/GPhilo/keras/tree/threading_bug_fix) on a possible fix for the shuffling that uses multiprocessing's `apply_async` function.
It works (at least for me on Windows) when `pickle_safe=False`, but I can't check the pickle-based implementation because I run in the "Can't pickle local object 'GeneratorEnqueuer.start.<locals>.data_generator_task" error mentioned in #5071 (any help testing the code on UNIX is very appreciated)

**Comment by Dref360** [comment]
One way to easily fix this would be implement a new API to look like the one from Pytorch. The generator would become indexed and the main thread would ask for input according to ids to idle processes. 
So the main threads ask for 1,2,3,4,5 to all 5 processes, the first one to return get the 6, then the 7. It's 1 more call between the main process and the others processes, but it would solve a lot of problem. Would be a major change though so likely to never happen

**Comment by GPhilo** [comment]
It's actually pretty easy to implement with `concurrent.futures` (or equivalently with `multiprocessing.pools.ThreadPool` and its method `apply_async()` which essentially does the same as the futures). I played around with both today and came up with some pretty good results limited, however, to threads.
From what I could see today, generators are not supposed to be sent around processes at all (since they're not picklable and Windows doesn't have `fork()`) and what the code does if `pickle_safe=True` is both platform-dependent and either buggy or broken.
I implemented the interface for the threads, but I'm stuck for the equivalent with multiprocessing... mostly because I'm not sure there should be at all an interface based on multiprocessing.

**Comment by Dref360** [comment]
For multiprocessing, you need some sort of indexing because the generator is copied.
Something like this would work.
https://gist.github.com/Dref360/30e0eda6f8f748460057ab2bb2aa835a

**Comment by GPhilo** [comment]
Uhm I think I'm missing something. I apologize for my long posts, but I'm trying to be as clear as possible explaining my reasoning because I'm not sure I got everything right.
- I am working under the assumption that the library should work equally under windows and linux, since both backends are available for both platforms (but I studied the behaviour under linux to see if at least there it works).
- I also assume that the specific order in which the generator produces the batches is important and should not be altered neither by reordering the batches, nor by doubling/cloning batches when these are enqueued. 

I think this is the expected behaviour because the only difference between generating the batches yourself and calling per-batch `{train, evaluate, predict}_batch` and building a generator and passing it to `{train, evaluate, predict}_generator` should IMO be the better resource usage of the generator that comes from queueing more batches so that the network can run as fast as it can. 

From my research and experiments, I have seen that generators are not picklable and thus cannot be passed to a subprocess that is spawned on windows. This already breaks my first assumptoin, but I'll put windows aside for a moment and look at what happens in Linux.

Due to the generation of processes via `fork()`, the generator is split into identical clones, each of which wll then produce *the same sequence of batches, **independently***. Let's call `bi_j` the `i`th batch generated by the `j`th copy of the generator. If we have `workers=2`, *even if we enforce that in each process the batches are enqueued in the ordered produced by the local copy of te generator*, we can have different layouts of the batches in the common queue.

For example, we might have `b1_1, b2_1, b3_1, b1_2, b2_2, b3_2, ...` or `b1_1, b1_2, b2_1, b2_2, ...` or any other mix of the two sequences `bi_1`, `bi_2`.

This, IMO, should be considered a bug because
1) the specific queue layout is nondeterministic even when the generator is.
2) the amount of times a specific batch is enqueued is also nondeterministic, depends on `workers` and randomply ranges from 1 to `workers`
3) the `steps` parameter (and its equivalents) of the `*_generator` functions *counts one step per batch extracted by the queue*. If, for example, we want to `evaluate_generator` on 5 batches and we have 5 workers, *it is possible that the evaluation actually runs only on the first batch, cloned 5 times*. Thus, the measured accuracy is wrong.

**Comment by GPhilo** [comment]
Any updates on this?

**Comment by Dref360** [comment]
In PR https://github.com/fchollet/keras/pull/6891

**Comment by jonilaserson** [comment]
My workaround for this was to use `GeneratorEnqueuer` for resource allocation and then `predict_on_batch`. I don't care about the ordering, because the generator yields (X, filename) and then I just take the X to predict_on_batch.

        enqueuer = GeneratorEnqueuer(batch_generator, pickle_safe=False)
        enqueuer.start(max_q_size=10, workers=4)
        while True:
            try:
                X, info = enqueuer.queue.get(1)
                Y = self.model.predict_on_batch(X)
                label_objects(info, Y) # can do this using apply_async for efficiency

            except Queue.Empty:
                enqueuer.stop()
                break

**Comment by joeyearsley** [comment]
@GPhilo is this fixed now?

**Comment by GPhilo** [comment]
The new dataset API from pr #6891 fixes the problem.

**Comment by joeyearsley** [comment]
Please can you close the issue?
